pkg/utils/disruption/disruption.go

```diff
@@ -0,0 +1,123 @@
+/*
+Copyright The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package disruption
+
+import (
+	"context"
+	"fmt"
+	"math"
+	"strconv"
+
+	"github.com/samber/lo"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/utils/clock"
+	"knative.dev/pkg/logging"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"sigs.k8s.io/karpenter/pkg/apis/v1beta1"
+	"sigs.k8s.io/karpenter/pkg/cloudprovider"
+	"sigs.k8s.io/karpenter/pkg/scheduling"
+)
+
+// lifetimeRemaining calculates the fraction of node lifetime remaining in the range [0.0, 1.0].  If the TTLSecondsUntilExpired
+// is non-zero, we use it to scale down the disruption costs of candidates that are going to expire.  Just after creation, the
+// disruption cost is highest, and it approaches zero as the node ages towards its expiration time.
+func LifetimeRemaining(clock clock.Clock, nodePool *v1beta1.NodePool, node *v1.Node) float64 {
```

Comment:
As an alternative encapsulation: nodePool.lifetimeRemaining(nodeClaim).

I think we probably want to define time w.r.t. NodeClaim and not node.

Reply:
Maybe I'm forgetting that conversation, but I actually think that reasoning about it in terms of the instance lifetime is probably the right thing to do here, rather than the Node. Don't we reason about expiration in terms of the NodeClaim for NodeClaim disruption status conditions?

Reply:
Yeah you're absolutely right, we do calculate it on the nodeclaim, not on the node. I'll change it here.

Reply:
On the encapsulation, I'm not sure the juice is worth the squeeze on making it a nodepool member function. It's pretty specific to disruption. I could maybe see an encapsulation making it a nodeclaim member function, but considering it's just 3 parameters, I'm thinking it's fine as it is.